### DISCUSSION TITLE
Multipart Form does not seem to work

### DISCUSSION THREAD

**Discussion by hhedin** [flawiddsouza/Restfox#241] [Q&A] [answered]
Hi!

I am currently testing Restfox as an alternative to postman. The tool works great in general, but I cannot for the life of me get Multipart form data to work. I have a couple of APIs that rely on this and this is stopping me from presenting Restfox as a reliable alternative.

The same works well with curl, below is pretty much equivalent to what I cannot get to work in Restfox:

`curl -v --request POST --url 'https://example.com/api/test?param=test' --header 'Authorization: Basic authStringBase64' --header 'content-type: multipart/form-data' --form file=@test.txt`

I can see a difference in the generated content-type header. But I am unsure if this has anything to do with the issue:

**curl**
`Content-Type: multipart/form-data; boundary=------------------------bcc23c7d57e7ad89`

**Restfox**
`Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryJr7dbcBLEBjAutRG`

The receiving end is a Spring boot application, and it works with both postman and curl. Has anyone gotten Multipart Form to work in Restfox?

For reference I am running Restfox as a container in Openshift, but I have the same issue if I run it on localhost.

**Comment by flawiddsouza** [comment]
Hi, can you check if it works correctly if you use one of the desktop applications: https://github.com/flawiddsouza/Restfox/releases/tag/v0.24.0? Wonder if the issue is isolated to the docker version.

**Reply by hhedin** [reply]
Hey, I tried the snap on ubuntu and it seems to work!

I wonder if we have caused this issue by ourselves. The container version does not give me the alternative to disable SSL verification under settings as the Snap version does. Would be great if I could disable it in the container version as well, because my SSL "fix" might have broken something:

**lawiddsouza/restfox:latest settings**
![Skärmbild från 2024-09-11 14-47-14](https://github.com/user-attachments/assets/584cbacf-8ee1-4b31-9937-2007154cddbd)


**Ubuntu Snap settings**
![Skärmbild från 2024-09-11 14-47-27](https://github.com/user-attachments/assets/4f666c4a-f7de-4268-bc29-733e4f987747)

To quickly solve this I have used flawiddsouza/restfox:latest in an image build to move the Restfox artifacts to an image where I know that node has all the certs it needs:

```
FROM flawiddsouza/restfox:latest AS build

FROM example.com/nodejs-20-minimal-ubi9
USER root
RUN mkdir -p /app/web-standalone/public
WORKDIR /app/web-standalone
COPY --from=build /app/web-standalone /app/web-standalone
COPY --from=build /app/web-standalone/public /app/web-standalone/public
USER 1001
CMD ["npm", "start"]
```
So I realize I have actually not been able to test this with the vanilla image due to self signed certificates.

**Reply by flawiddsouza** [reply]
Let me see if I can add an option to disable ssl checking inside the docker container. That should fix your problem.

Multipart file upload might be bugged in the docker version, so let me also check that.

**Reply by flawiddsouza** [reply]
Hi @hhedin, new version released: https://github.com/flawiddsouza/Restfox/releases/tag/v0.25.0. This has both fix for multipart and option to disable ssl verification. Do let me know if something seems broken.

**Reply by hhedin** [reply]
Great work @flawiddsouza, I did a small test and it seems to work now!

I will test with some larger files during the day and mark the thread as answered if nothing comes up.

Thank you for the quick support!

**Reply by hhedin** [reply]
@flawiddsouza smaller text based files seems to work without issue in multipart. But I cannot get for example a .zip with random content (around 15mb) to work reliably in multipart.

I also had issues running the original image in kubernetes/openshift, this is likely due to kubernetes/openshift not wanting to run the image as root.

```
> restfox-web-standalone@0.0.0 start
> node app.js
npm ERR! code EACCES
npm ERR! syscall mkdir
npm ERR! path /.npm
npm ERR! errno -13
npm ERR!
npm ERR! Your cache folder contains root-owned files, due to a bug in
npm ERR! previous versions of npm which has since been addressed.
npm ERR!
npm ERR! To permanently fix this problem, please run:
npm ERR! sudo chown -R 1008670000:0 "/.npm"
npm ERR! Log files were not written due to an error writing to the directory: /.npm/_logs
npm ERR! You can rerun the command with `--loglevel=verbose` to see the logs in your terminal

```

**Reply by flawiddsouza** [reply]
Let me look into this. The larger upload size issue might be a simple fix. Not sure what to do about the 2nd issue. A bit confused on why the output error is from npm rather than node.

**Reply by hhedin** [reply]
For the second issue I overlooked that the error was from npm, strange as you say! The image does not show the same behaviour on my local machine with neither podman or docker, so it is likely something on my end.

**Reply by flawiddsouza** [reply, marked as answer]
Hi @hhedin, I multipart uploaded a 17 MB exe file & 51 MB zip file multiple times through the the docker run version of Restfox and faced no issues with the upload. Can you check once locally by running in podman or docker? To isolate the problem from your side. Maybe the problem is arising at kubernetes/openshift.

**Reply by hhedin** [reply]
@flawiddsouza It seems like Restfox in Openshift and Restfox with Podman works pretty much the same. I experimented a bit today and smaller files worked fine for both, the Issue I had seems to be related to memory consumption and restriction during the upload.

I hit a limit in Openshift around 200 MB upload I believe, the container hits the configured memory limit of around 2,5 GB during the upload. Working towards localhost (docker/podman run) I can upload larger files, I stopped arount 600 MB and the container consumed around 3,6 GB I believe. I tested our most extreme test case that we are migrating from Postman which involves a file of around 2.4 GB of data, I could not get this to work on localhost either.

I think that the current functionality of Restfox with file upload will satisfy most users, we just have a special testcase here. We will use curl for these file heavy API tests and Restfox for the other stuff.

I will mark as resolved as we got multipart to work for most part, thanks for the support!

**Reply by flawiddsouza** [reply]
When I have time, I'll be testing and checking if there's a way to get the docker version to handle larger files without consuming too much memory. Will message here if I find a way.
